### PR TITLE
fix: thread-safe improvements

### PR DESCRIPTION
**Pull request by qingzhuozhen**
### Summary

- Fix issues for multi-threading case
- Set up some unit tests to help catch possible issues.
- Gathering thoughts or suggestions on the fixes or potential issues.

### Checklist

* [ ] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-Java/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:

### CODE REVIEW

File: src/main/java/com/amplitude/HttpTransport.java [hunk 1]
```diff
@@ -161,11 +150,11 @@ private void retryEventsOnLoop(String userId, String deviceId) {
                   }
                   if (shouldReduceEventCount) {
                     if (!isLastTry) {
-                      eventCount /= 2;
```
Comment (justin-fiedler) [inline src/main/java/com/amplitude/HttpTransport.java:164]: Why'd you move this down?

File: src/test/java/com/amplitude/AmplitudeMultiThreadTest.java [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    CountDownLatch threadLatch = new CountDownLatch(threads);
+    ExecutorService executorService = Executors.newFixedThreadPool(threads);
+
+    for (int i = 0; i < threads; i++) {
+      executorService.execute(new LogEventRunnable(amplitude, total / threads, threadLatch));
+    }
+
+    try {
+      threadLatch.await();
+      callbacks.blockWithTimeout(15);
```
Comment (justin-fiedler) [inline src/test/java/com/amplitude/AmplitudeMultiThreadTest.java:63]: Why do you need this timeout?

File: src/main/java/com/amplitude/HttpTransport.java (hunk 1 above)
Comment (qingzhuozhen) [reply src/main/java/com/amplitude/HttpTransport.java:164]: I realized the original way is not correct. Basically eventsBuffer is not changed in the loop or by subList. The original way we just call more event callbacks then we want to.

File: src/test/java/com/amplitude/AmplitudeMultiThreadTest.java (hunk 2 above)
Comment (qingzhuozhen) [reply src/test/java/com/amplitude/AmplitudeMultiThreadTest.java:63]: We need this timeout so when there is any issues the tests won't just hang there..

File: src/main/java/com/amplitude/Amplitude.java [hunk 3]
```diff
@@ -107,7 +108,7 @@ public void setCallbacks(AmplitudeCallbacks callbacks) {
    *
    * @param event The event to be sent
    */
-  public void logEvent(Event event) {
+  public synchronized void logEvent(Event event) {
```
Comment (qingzhuozhen) [inline src/main/java/com/amplitude/Amplitude.java:111]: The goal for this is between add and clear we don't lose events..

File: src/main/java/com/amplitude/HttpTransport.java [hunk 4]
```diff
@@ -308,9 +292,49 @@ private void triggerEventCallbacks(List<Event> events, int status, String messag
     if (callbacks == null || events == null || events.isEmpty()) {
       return;
     }
-
     for (Event event : events) {
       callbacks.onLogEventServerResponse(event, status, message);
     }
   }
+
+  private void addEventToBuffer(String userId, String deviceId, Event event) {
```
Comment (qingzhuozhen) [inline src/main/java/com/amplitude/HttpTransport.java:300]: The goal here is for add or remove we don't lose events on the buffer.

File: src/main/java/com/amplitude/HttpTransport.java [hunk 5]
```diff
@@ -231,7 +220,7 @@ private EventsRetryResult retryEventsOnce(String userId, String deviceId, List<E
           || events.size() == 1) {
         // Return early if there's an issue with the entire payload
         // or if there's only one event and its invalid
-        triggerEventCallbacks(eventsToDrop, response.code, response.error);
+        triggerEventCallbacks(events, response.code, response.error);
```
Comment (qingzhuozhen) [inline src/main/java/com/amplitude/HttpTransport.java:223]: Covers some callbacks weren't triggered

File: src/test/java/com/amplitude/AmplitudeMultiThreadTest.java (hunk 2 above)
Comment (justin-fiedler) [reply src/test/java/com/amplitude/AmplitudeMultiThreadTest.java:63]: Oh got it. I thought it was an additional wait, not a timeout. Makes sense.

**Review by justin-fiedler — APPROVED** [review]

**Pull request merged**